go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -266,9 +398,25 @@ func (c *Controller) syncService(key string) error {
 			},
 		}
 
-		if err := ovnlb.EnsureLBs(c.nbClient, service, nil); err != nil {
-			return fmt.Errorf("failed to delete load balancers for service %s/%s: %w",
-				namespace, name, err)
+		c.alreadyAppliedLock.Lock()
+		existingLBs, ok := c.alreadyApplied[key]
+		c.alreadyAppliedLock.Unlock()
+
+		if ok {
+			//
+			// The controller's alreadyApplied functions as the cache for the service controller to map into OVN
+			// load balancers. While EnsureLBs may be concurrently called by this controller's workers, only a single
+			// worker will be operating at a given service. That is why it is safe to have changes to this cache
+			// from multiple workers, because the `key` is always uniquely hashed to the same worker thread.
+
+			if err := ovnlb.EnsureLBs(c.nbClient, service, existingLBs, nil); err != nil {
```

Comment:
isn't this technically unsafe? You are passing the slice by reference, which will modify the alreadyApplied cache outside of the lock:

https://go.dev/play/p/a3uRFoa5uqZ

Reply:
You make a good point. In my `defense`, that behavior predates the changes in this PR.
Still, this may be a good opportunity to fix it.
I will modify EnsureLBs to return the changes needed (instead of doing them in there)  and have this caller do them for real after grabbing the lock. That should address this concern, agree?

Reply:
It looks like EnsureLBs does not actually modify the contents, it only reads...albeit an unsafe read. You need to copy the slice and then pass it to EnsureLBs:

https://go.dev/play/p/7f5T5c8hJwA

Reply:
We modify LBs inside  `EnsureLBs` here: 

https://github.com/ovn-org/ovn-kubernetes/blob/4bb783f48b9e7579f4e6a132986bb4c054832725/go-controller/pkg/ovn/loadbalancer/loadbalancer.go#L201-L203

So we definitely could potentially get in trouble with goLang because that is ensureLbs w/out holding any locks. Copying it would not be good, unless we copied it back, which feels `expensive`.

Reply:
Nit: using `copy(dst, src)` instead of 
```
	for idx, val := range existingLBs {
		copyLBs[idx] = val
	}
```
That is ok, right?

Reply:
oh yeah, sure. I forgot about that one.